signer/core/apiv2.go

```diff
@@ -0,0 +1,259 @@
+	}
+
+	for typeKey, typeArr := range types {
+		var typeValArr []string
+
+		for _, typeObj := range typeArr {
+			typeVal := typeObj["type"]
+			if typeKey == typeVal {
+				return "", fmt.Errorf("type %s cannot reference itself", typeVal)
+			}
+
+			firstChar := []rune(typeVal)[0]
+			if unicode.IsUpper(firstChar) {
+				if types[typeVal] != nil {
+					if !visited(typeValArr, typeVal) {
+						typeValArr = append(typeValArr, typeVal)
+						update(typeKey, typeVal)
+					}
+				} else {
+					return "", fmt.Errorf("referenced type %s is undefined", typeVal)
+				}
+			} else {
+				if !types.IsStandardType(typeVal) {
+					if types[typeVal] != nil {
+						return "", fmt.Errorf("Custom type %s must be capitalized", typeVal)
+					} else {
+						return "", fmt.Errorf("Unknown type %s", typeVal)
+					}
+				}
+			}
+		}
+
+		typeValArr = []string{}
+	}
+
+	sortedPriorities := types.SortByPriorityAndName(priorities)
+	var buffer bytes.Buffer
+	for _, priority := range sortedPriorities {
+		typeKey := priority.Type
+		typeArr := types[typeKey]
+
+		buffer.WriteString(typeKey)
+		buffer.WriteString("(")
+
+		for _, typeObj := range typeArr {
+			buffer.WriteString(typeObj["type"])
+			buffer.WriteString(" ")
+			buffer.WriteString(typeObj["name"])
+			buffer.WriteString(",")
+		}
+
+		buffer.Truncate(buffer.Len() - 1)
+		buffer.WriteString(")")
+	}
+
+	return buffer.String(), nil
+}
+
+func encodeData(values EIP712Message) (string, error) {
+	return "", nil
+}
+
+// Checks if the given type is a standard type accepted by EIP-712
+func (types *EIP712Types) IsStandardType(typeStr string) bool {
+	standardTypes := []string{
+		"array",
+		"address",
+		"boolean",
+		"bytes",
+		"string",
+		"struct",
+		"uint",
+	}
+	for _, val := range standardTypes {
+		if strings.HasPrefix(typeStr, val) {
+			return true
+		}
+	}
+	return false
+}
+
+// Helper function to sort types by priority and name. Priority is calculated b
+// based upon the number of references.
+func (types *EIP712Types) SortByPriorityAndName(input map[string]uint) []EIP712TypePriority {
+	var priorities []EIP712TypePriority
+	for key, val := range input {
+		priorities = append(priorities, EIP712TypePriority{key, val})
+	}
+	// Alphabetically
+	sort.Slice(priorities, func(i, j int) bool {
+		return priorities[i].Type < priorities[j].Type
+	})
+	// Priority
+	sort.Slice(priorities, func(i, j int) bool {
+		return priorities[i].Value > priorities[j].Value
+	})
+
+	for _, priority := range priorities {
+		fmt.Printf("%s, Value %d\n", priority.Type, priority.Value)
+	}
+	fmt.Printf("\n")
+
+	return priorities
+}
+
+// Check if the given domain is valid, i.e. contains at least the minimum viable keys and values
+func (domain *EIP712Domain) IsValid() error {
+	if domain.ChainId == big.NewInt(0) {
+		return fmt.Errorf("chainId must be specified according to EIP-155")
+	}
+
+	if domain.Name == "" && domain.Version == "" && len(domain.VerifyingContract) == 0 && len(domain.Salt) == 0 {
```

Comment:
in general, and it perhaps doesn't matter here, but using `len(domain.Name) == 0` is more robust, since it is true also if the thing you're testing is `nil`.